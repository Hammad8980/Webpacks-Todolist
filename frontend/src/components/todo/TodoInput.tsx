import { useState, type FormEvent } from 'react';
import { Task } from '../../features/todos/TodoTaskTypes';
import Button from '../ui/Button';

type TodoInputProps = {
  tasks: Task[];
  onAddTask: (taskName: string) => void;
};

function TodoInput({ tasks, onAddTask }: TodoInputProps) {
  const [taskName, setTaskName] = useState('');
  const [error, setError] = useState('');

  const handleAddTasks = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault(); // This is to prevent the browser from refreshing the page

    setError('');

    if (!taskName.trim()) {
      setError('Please enter a task');
      return;
    }

    const isDuplicate = tasks.some(
      task => task.title.toLowerCase().trim() === taskName.toLowerCase().trim()
    );

    if (isDuplicate) {
      setError('Task already exists!');
      setTaskName(''); // Clear input field
      setTimeout(() => setError(''), 2500);
      return;
    }
    onAddTask(taskName);
    setTaskName('');
  };
  return (
    <div className="space-y-2">
      <form className="flex" onSubmit={handleAddTasks}>
        <input
          type="text"
          value={taskName}
          placeholder="What needs to be done?"
          onChange={e => setTaskName(e.target.value)}
          className="flex-grow p-3 bg-gray-700 border border-gray-600 rounded-l-lg text-white placeholder-gray-400 focus:outline-none focus:ring-0 focus:ring-inset focus:border-blue-500"
        />

        <Button
          className="bg-gradient-to-r from-blue-600 to-blue-700 hover:from-blue-700 hover:to-blue-800 text-white font-semibold py-3 px-6 rounded-r-lg transition-all duration-200 shadow-lg hover:shadow-xl"
          type="submit"
        >
          <svg className="w-5 h-5" fill="none" stroke="currentColor" viewBox="0 0 24 24">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 4v16m8-8H4" />
          </svg>
        </Button>
      </form>

      {/* Error message display */}
      {error && (
        <div className="bg-red-100 border border-red-400 text-red-700 px-4 py-3 rounded-lg shadow-md animate-fade-in">
          <div className="flex items-center">
            <svg className="w-5 h-5 mr-2" fill="currentColor" viewBox="0 0 20 20">
              <path
                fillRule="evenodd"
                d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
                clipRule="evenodd"
              />
            </svg>
            <span className="font-medium">{error}</span>
          </div>
        </div>
      )}
    </div>
  );
}

export default TodoInput;
