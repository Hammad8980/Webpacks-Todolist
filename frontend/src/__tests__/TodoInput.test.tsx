import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoInput from '../components/todo/TodoInput';
import type { Task } from '../features/todos/TodoTaskTypes';

describe('TodoInput', () => {
  const mockOnAddTask = jest.fn();

  const defaultProps = {
    tasks: [] as Task[],
    onAddTask: mockOnAddTask,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders input field and submit button', () => {
    const { container } = render(<TodoInput {...defaultProps} />);

    expect(screen.getByPlaceholderText('What needs to be done?')).toBeInTheDocument();
    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('calls onAddTask when form is submitted using click or enter with valid input', async () => {
    const user = userEvent.setup();
    render(<TodoInput {...defaultProps} />);

    const input = screen.getByPlaceholderText('What needs to be done?');
    const submitButton = screen.getByRole('button');

    await user.type(input, 'New task');
    await user.click(submitButton);

    await user.type(input, 'New task{enter}');

    expect(mockOnAddTask).toHaveBeenCalledWith('New task');
    expect(mockOnAddTask).toHaveBeenCalledTimes(2);
    expect(input).toHaveValue('');
  });

  it('shows error message when trying to submit empty task', async () => {
    const user = userEvent.setup();
    const { container } = render(<TodoInput {...defaultProps} />);

    const submitButton = screen.getByRole('button');
    await user.click(submitButton);

    expect(screen.getByText('Please enter a task')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    expect(mockOnAddTask).not.toHaveBeenCalled();
  });

  it('shows error message when trying to add duplicate task', async () => {
    const user = userEvent.setup();
    const existingTasks: Task[] = [
      { id: 1, title: 'Existing task', isCompleted: false, priority: 'p2' },
    ];

    render(<TodoInput {...defaultProps} tasks={existingTasks} />);

    const input = screen.getByPlaceholderText('What needs to be done?');
    const submitButton = screen.getByRole('button');

    await user.type(input, 'Existing task');
    await user.click(submitButton);

    expect(screen.getByText('Task already exists!')).toBeInTheDocument();
    expect(input).toHaveValue('');
    expect(mockOnAddTask).not.toHaveBeenCalled();
  });

  it('error message disappears after timeout', async () => {
    jest.useFakeTimers();
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

    const existingTasks: Task[] = [
      { id: 1, title: 'Existing task', isCompleted: false, priority: 'p2' },
    ];

    render(<TodoInput {...defaultProps} tasks={existingTasks} />);

    const input = screen.getByPlaceholderText('What needs to be done?');
    const submitButton = screen.getByRole('button');

    await user.type(input, 'Existing task');
    await user.click(submitButton);

    expect(screen.getByText('Task already exists!')).toBeInTheDocument();

    jest.advanceTimersByTime(2500);

    await waitFor(() => {
      expect(screen.queryByText('Task already exists!')).not.toBeInTheDocument();
    });

    jest.useRealTimers();
  });
});
