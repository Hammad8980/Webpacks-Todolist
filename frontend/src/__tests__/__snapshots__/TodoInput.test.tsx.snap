// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoInput renders input field and submit button 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <form
      class="flex"
    >
      <input
        class="flex-grow p-3 bg-gray-700 border border-gray-600 rounded-l-lg text-white placeholder-gray-400 focus:outline-none focus:ring-0 focus:ring-inset focus:border-blue-500"
        placeholder="What needs to be done?"
        type="text"
        value=""
      />
      <button
        class="bg-gradient-to-r from-blue-600 to-blue-700 hover:from-blue-700 hover:to-blue-800 text-white font-semibold py-3 px-6 rounded-r-lg transition-all duration-200 shadow-lg hover:shadow-xl"
        type="submit"
      >
        <svg
          class="w-5 h-5"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4v16m8-8H4"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;

exports[`TodoInput shows error message when trying to submit empty task 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <form
      class="flex"
    >
      <input
        class="flex-grow p-3 bg-gray-700 border border-gray-600 rounded-l-lg text-white placeholder-gray-400 focus:outline-none focus:ring-0 focus:ring-inset focus:border-blue-500"
        placeholder="What needs to be done?"
        type="text"
        value=""
      />
      <button
        class="bg-gradient-to-r from-blue-600 to-blue-700 hover:from-blue-700 hover:to-blue-800 text-white font-semibold py-3 px-6 rounded-r-lg transition-all duration-200 shadow-lg hover:shadow-xl"
        type="submit"
      >
        <svg
          class="w-5 h-5"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 4v16m8-8H4"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </form>
    <div
      class="bg-red-100 border border-red-400 text-red-700 px-4 py-3 rounded-lg shadow-md animate-fade-in"
    >
      <div
        class="flex items-center"
      >
        <svg
          class="w-5 h-5 mr-2"
          fill="currentColor"
          viewBox="0 0 20 20"
        >
          <path
            clip-rule="evenodd"
            d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7 4a1 1 0 11-2 0 1 1 0 012 0zm-1-9a1 1 0 00-1 1v4a1 1 0 102 0V6a1 1 0 00-1-1z"
            fill-rule="evenodd"
          />
        </svg>
        <span
          class="font-medium"
        >
          Please enter a task
        </span>
      </div>
    </div>
  </div>
</div>
`;
